dali/pipeline/data/buffer.cc

```diff
@@ -18,22 +18,48 @@
 
 namespace dali {
 
-// this is to make debug builds happy about kMaxGrowthFactor
-template class Buffer<CPUBackend>;
-template class Buffer<GPUBackend>;
+DLL_PUBLIC AccessOrder get_deletion_order(const std::shared_ptr<void> &ptr) {
+  if (auto *del = std::get_deleter<mm::AsyncDeleter>(ptr))
+    return AccessOrder(del->release_on_stream);
+  else
+    return {};
+}
+
+DLL_PUBLIC bool set_deletion_order(const std::shared_ptr<void> &ptr, AccessOrder order) {
+  if (ptr.use_count() == 1 && order.has_value()) {
+    if (auto *del = std::get_deleter<mm::AsyncDeleter>(ptr)) {
+      del->release_on_stream = order.get();
+      if (ptr.use_count() != 1)
```

Comment:
What if it changed to 2 and then returned to 1 between L31 and L32?

Reply:
Then we're good. The program would be still "correct" (i.e. it wouldn't crash) if we set deletion order every time, but the performance would be affected. Buffer objects are not thread-safe anyway, so if we detected such a problem, then it would indicate a more serious issue.

Reply:
So why do we throw here if this is just a performance optimization?

Reply:
Note that this is a re-check - if someone has somehow increased the use count of this pointer, then something very wrong is happening. This _could_ be skipped, but I prefer to keep this check here just in case. After all, when we're _not_ throwing, it's just one extra if.

Reply:
Understood.